contributors/design-proposals/network/pod-resolv-conf.md

```diff
@@ -0,0 +1,167 @@
+# Custom /etc/resolv.conf
+* Status: pending
+* Version: alpha
+* Implementation owner: Bowei Du <[bowei@google.com](mailto:bowei@google.com)>, Zihong Zheng <[zihongz@google.com](mailto:zihongz@google.com)>
+
+# Overview
+The `/etc/resolv.conf` in a pod is managed by Kubelet and its contents are generated based on `pod.dnsPolicy`. For `dnsPolicy: Default`, `resolv.conf` is copied from the node where the pod is running. If the `dnsPolicy` is `ClusterFirst`, the contents of the resolv.conf is the hosts `resolv.conf` augmented with the following options:
+
+*   Search paths to add aliases for domain names in the same namespace and cluster suffix.
+*   `options ndots` to 5 to ensure the search paths are searched for all potential matches.
+
+The configuration of both search paths and `ndots` results in query amplification of five to ten times for non-cluster internal names. This is due to the fact that each of the search path expansions must be tried before the actual result is found. This order of magnitude increase of query rate imposes a large load on the kube-dns service. At the same time, there are user applications do not need the convenience of the name aliases and do not wish to pay this performance cost.
+
+
+## Existing workarounds
+
+The current work around for this problem is to specify an FQDN for name resolution. Any domain name that ends with a period (e.g. `foo.bar.com.`) will not be search path expanded. However, use of FQDNs is not well-known practice and imposes application-level changes. Cluster operators may not have the luxury of enforcing such a change to applications that run on their infrastructure.
+
+It is also possible for the user to insert a short shell script snippet that rewrites `resolv.conf` on container start-up. This has the same problems as the previous approach and is also awkward for the user. This also forces the container to have additional executable code such as a shell or scripting engine which increases the applications security surface area.
+
+
+# Proposal sketch
+
+Add a new `pod.dnsPolicy: Custom` that allows for user customization of `resolv.conf`.
+
+
+## Pod API example
+
+In the example below, the user wishes to add the pod namespace and a custom expansion to the search path, as they do not use the other name aliases:
+
+```yaml
+# Pod spec
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: ns1
+  name: example
+spec:
+  containers:
+  - name: example
+    image: example
+  dnsPolicy: Custom
+  dnsParams:
+    searchPaths:
+    - $NAMESPACE.svc.$CLUSTER
```

Comment:
We should be very careful with templating syntax.  This could set precedent.

I would suggest $(NAMESPACE) similar to other places we do expansion.  Needs REALLY good docs on what expansions are available and what happens if the user uses an invalid one, and how to get a literal value from the escape.

Reply:
Maybe you can get away without templating, at least initially? My guess is that any user who needs such an advanced feature can figure out how to template this themselves, or already have a templating layer anyway (we do).

This would uncouple the DNS configurability from all the complexity of adding templating.

If it turns out to be really indispensable, it can still be added later, but once it's in you can't get it out and you'll have to explain why it's not available everywhere.

Reply:
I added more detail to the psuedocode and exposition. The system works by substituting labels if the substitution value matches and is not intended to be a generic template mechanism.

@matthiasr we can leave out templating but suspect most people would want to be able to write YAMLs portable across namespaces without forcing use of another layer.